### ISSUE TITLE
[Discovery] Can we use GraphQL to generate articles list?

### ISSUE THREAD

**Issue by MickinSahni** [department-of-veterans-affairs/va.gov-team#13883]
## Issue Description
Given we need to generate lists of articles in the LC and we have a potential approach from the CMS team that leverages GraphQL, I want to vet that we can use this approach to return data required to generate static pages for taxonomy terms aka "articles tagged by".

### use cases
- person adds a new tag
- person adds more articles associated with a new or existing tag

---
## Tasks
- [ ] Document findings here
- [ ] Communicate with the CMS team

## Acceptance Criteria
- We've determined feasibility of the GraphQL approach to generate list of articles in the LC

## Questions
- The biggest complication/question is how Transformers will affect this approach. Right now every query represents a particular page. In this case, it requires a more sophisticated query. BUT, do we really need to account for Transformers *right now*? The reason being we'd always need this GraphQL implementation for previewing nodes. 
  - **tldr: the Transformers concern doesn't prevent us from using this GraphQL approach since we'll need it for previewing**
- Besides Transformers, our previews use node ids and if this isn't a node i.e. it's an index page, how will the preview server show it? 
- How does the CMS Demo Environment impact using GraphQL to generate articles list (if at all)?

--- 

<img width="977" alt="image" src="https://user-images.githubusercontent.com/13204473/93935840-400bf780-fcf3-11ea-868d-fb8b7ba2463f.png">

**Comment by MickinSahni** [comment]
Concerns
- how is this affected by transformers

**Comment by kelsonic** [comment]
Since we've discovered this is doable with GraphQL, can we close this?

**Comment by ncksllvn** [comment]
So after thinking about this, I realized we don't really need any additional GraphQL - we're already using GraphQL to query for all of the pages that will be listed under the articles list, so we can just derive the article list off of that. I think that's our best option, because it won't be affected by transformers.

My one concern is previewing this. The preview server operates on a nodeId. Since the article listing will be generated and not powered by a nodeId, there isn't a clear mechanism for how preview would work. In fact, the more I think about it, I'm not sure how we can make preview work at all for this...A CMS demo environment would do the trick but that's not super ideal post-launch because we may not want to keep a demo environment maintained

**Comment by MickinSahni** [comment]
> we're already using GraphQL to query for all of the pages that will be listed under the articles list, so we can just derive the article list off of that

so that sounds good. but does that change if we can't preview it? 

> I'm not sure how we can make preview work at all for this

"all of this" being all LC templates or ...?

**Comment by ncksllvn** [comment]
> so that sounds good. but does that change if we can't preview it?

Without a Drupal nodeId, I don't see any way for us to build the article listings where we can preview it

> "all of this" being all LC templates or ...?

"all of this" being specifically the article listings

**Comment by MickinSahni** [comment]
Hmm perhaps we can take the CMS demo env idea to our CMS sync this afternoon to see what CMS folks might think. 

Sounds like it's the only option at this point.

**Comment by oksana-c** [comment]
Content listings are powered by Taxonomy terms, which have TIDs (Term ID) similar to node ID. Taxonomy terms will also have URL aliases (alias patterns TBD), which should be used to generate corresponding pages on FE.
Let's touch base on preview strategy for Drupal taxonomy terms at 4pm today - added to agenda.

**Comment by ncksllvn** [comment]
TIL TIDS! Sounds good

**Comment by MickinSahni** [comment]
* Have we powered a preview using TID? No, but it's the same concept as using node ids
* How about using this approach on the va.gov homepage? good question. unclear without more background
* so we can build preview pages based on the tags tailored to LC needs first, before optimizing for future va.gov products
* Do we need to preview? It's useful to increase our confidnece 
  - LoE of using TID to preview: not difficult to use TID instead of NodeID. But, what does this mean to our data structure we rely on? It means we have to query for all articles tagged with that TID. 
    - but where do we get the TID from to know what to call?

**Comment by jenniferlee-dsva** [comment]
Continuing discussion here - thanks @MickinSahni for capturing the summary. 

My 2 cents is:  

- For content pages like articles and benefit detail pages, preview is an absolutely needed functionality.
- For pages like this where we're not per se manually adding/editing content -- but rather the page is assembling content data and putting it together -- preview would be a nice to have. 
- I'm not super convinced about the benefit here though of previewing. (For example, if an author adds a new article with tag X, and they preview the list page, they will likely need to paginate (possibly alot) to validate that the article was added and in the correct alpha order, and showing the content type label.)  From an authoring exp, this doesn't seem that useful  - it's just a list and if I can validate it w/in the CMS, that seems do-able. 

HOWEVER - might there be other benefits to creating a preview of said list?

**Comment by DanielleThierryUSDSVA** [comment]
Hi @jenniferlee-dsva. Agree 100% that preview is a necessary for content pages.

For things like link lists, I agree that preview isn't essential. My only thought is that being able to fully visualize how an article fits into the larger context of content presentation might help authors better assess the need for and relevance of their new article.

**Comment by ncksllvn** [comment]
Hm, okay, so article listing previews is at least not an absolute requirement right now, although it would be nice to have. I think @oksana-c 's solution proposed using TIDs is probably the most correct way of doing this, but my concerns are -

1. How do transformers transform a GraphQL query involving TIDs? 
1. How does the website build/deployment work with TIDs? (The preview server would receive a "TID" as a request parameter, but the website build does not. How does the website build know the TID keys/values?)

However, the quickest and most straightforward way to implement this would likely be what I proposed before - derive the article listings page without a separate GraphQL query. This way, there is no concern of a rewrite due to transformer impact (because we aren't writing any new GraphQL), but there is no way to support the preview server with this approach.

**Comment by MickinSahni** [comment]
Decision: we'll validate articles list within CMS, not on preview env or va.gov. This allows us to validate the logic of the query.

**Comment by jenniferlee-dsva** [comment]
Plus Mickin's comment ^ - For launch, we still want to make sure the FE is building these 'static' list pages okay during demo environment validation (as live-demo pages).
